augur/curate/apply_date_bounds.py

```diff
@@ -0,0 +1,231 @@
+
+    failures = []
+
+    for index, input_record in enumerate(records):
+        record = input_record.copy()
+        try:
+            record[args.date_field] = Record(record, index).get_bounded_date(args)
+        except DataError as error:
+            if args.failure_reporting is DataErrorMethod.SILENT:
+                continue
+            if args.failure_reporting is DataErrorMethod.ERROR_FIRST:
+                raise error
+            if args.failure_reporting is DataErrorMethod.WARN:
+                print_err(f"WARNING: {error}")
+                failures.append(error)
+                continue
+            if args.failure_reporting is DataErrorMethod.ERROR_ALL:
+                failures.append(error)
+                continue
+            else:
+                raise ValueError(f"Encountered unhandled failure reporting method: {args.failure_reporting!r}")
+        yield record
+
+    if args.failure_reporting is not DataErrorMethod.SILENT and failures:
+        if args.failure_reporting is DataErrorMethod.ERROR_ALL:
+            raise AugurError(dedent(f"""\
+                Unable to apply bounds. All errors:
+                {indented_list(map(str, failures), "                ")}"""))
+
+
+def validate_arguments(args: argparse.Namespace):
+    if not args.lower_bound and not args.upper_bound:
+        raise AugurError("At least one of --lower-bound and --upper-bound is required.")
+
+
+class Record:
+    """
+    Helper class to wrap a record, its id, and arguments for ease of error handling and small functions.
+    """
+
+    def __init__(self, data: RecordType, record_id: int) -> None:
+        self.data = data
+        self.id = record_id
+
+    def get_bounded_date(self, args: argparse.Namespace) -> str:
+        """
+        Returns a date string representing the date with bounds applied. bounded interval for the given record.
+        """
+        start, end = self.convert_date_to_range(args.date_field)
+
+        lower_bound, upper_bound = self.get_bounds(args.lower_bound, args.upper_bound)
+
+        # If any ends are unbounded, return the original date
+        if (start == float("-inf") and lower_bound is None) or \
+           (end   == float( "inf") and upper_bound is None):
+            return self.data[args.date_field]
+
+        # Error if start or end are out of bounds.
+        if lower_bound and start < lower_bound and end < lower_bound:
+            self.raise_data_error(
+                f"{args.date_field!r}={self.data[args.date_field]!r} "
+                f"is earlier than the lower bound of "
+                f"{args.lower_bound!r}={self.data[args.lower_bound]!r}"
+            )
+        if upper_bound and start > upper_bound and end > upper_bound:
+            self.raise_data_error(
+                f"{args.date_field!r}={self.data[args.date_field]!r} "
+                f"is later than the upper bound of "
+                f"{args.upper_bound!r}={self.data[args.upper_bound]!r}"
+            )
```

Comment:
`self.data[args.lower_bound]` would lead to a `KeyError` if the `args.lower_bound` was just a date string (same for the `args.upper_bound`.